plugins/basic/nexus-repository-docker/src/main/java/org/sonatype/nexus/repository/docker/internal/DockerFacetImpl.java

```diff
@@ -0,0 +1,669 @@
+   * Find all components (tags) given a component name.
+   *
+   * @param tx   storage transaction
+   * @param name image name
+   * @return found components
+   */
+  private Iterable<Component> findComponents(final StorageTx tx, final String name) {
+    ComponentQuery query = new Builder()
+        .where(P_NAME + "=").param(name)
+        .build();
+    return tx.findComponents(
+        query.getWhere(), query.getParameters(), singletonList(getRepository()), null
+    );
+  }
+
+  /**
+   * Find an asset by its name.
+   *
+   * @param tx        storage transaction
+   * @param bucket    repository bucket
+   * @param assetName asset name
+   * @return found asset or null if not found
+   */
+  @Nullable
+  private Asset findAsset(final StorageTx tx, final Bucket bucket, final String assetName) {
+    return tx.findAssetWithProperty(StorageFacet.P_NAME, assetName, bucket);
+  }
+
+  /**
+   * Validate that layer metadata is a proper JSON map.
+   *
+   * @param layerMetadataContent layer metadata content
+   * @return parsed layer metadata
+   */
+  private Map<String, Object> readLayerMetadata(final Supplier<InputStream> layerMetadataContent) {
+    TypeReference<Map<String, Object>> typeRef = new TypeReference<Map<String, Object>>() {};
+    try {
+      return mapper.readValue(layerMetadataContent.get(), typeRef);
+    }
+    catch (Exception e) {
+      throw new V1Exception("Invalid JSON");
+    }
+  }
+
+  /**
+   * Validate that images index is a proper JSON list containing layer references.
+   *
+   * @param imagesIndexContent images index content
+   * @return parsed images index
+   */
+  private List<Map<String, Object>> readImagesIndex(final Supplier<InputStream> imagesIndexContent) throws IOException {
+    try (InputStream in = imagesIndexContent.get()) {
+      TypeReference<List<Map<String, Object>>> typeRef = new TypeReference<List<Map<String, Object>>>() {};
+      try {
+        return mapper.readValue(in, typeRef);
+      }
+      catch (Exception e) {
+        throw new V1Exception("Invalid JSON");
+      }
+    }
+  }
+
+  /**
+   * Validate layer id.
+   *
+   * @param layerId to validate
+   * @return validated layer id
+   */
+  private String validateLayerId(final String layerId) {
+    if (!validLayerId.matcher(layerId).matches()) {
+      throw new V1Exception("Invalid image ID", HttpStatus.NOT_FOUND);
+    }
+    return layerId;
+  }
+
+  /**
+   * Validate tag name.
+   *
+   * @param tag to validate
+   */
+  private void validateTagName(final String tag) {
+    if (!validTag.matcher(tag).matches()) {
+      throw new V1Exception("Invalid tag name (must match ${VALID_TAG})");
+    }
+  }
+
+  /**
+   * Read layer id.
+   *
+   * @param streamSupplier layer id content
+   * @return parsed layer id
+   */
+  private String readLayerId(final Supplier<InputStream> streamSupplier) {
+    try {
+      return validateLayerId(mapper.readValue(streamSupplier.get(), String.class));
+    }
+    catch (Exception e) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1166)